helix-core/src/main/java/org/apache/helix/task/TaskUtil.java

```diff
@@ -1142,6 +1142,35 @@ public static void workflowGarbageCollection(final Set<String> toBePurgedWorkflo
     }
   }
 
+  /**
+   * The function that removes IdealStates and job contexts of the jobs that need to be
+   * deleted.
+   * @param toBePurgedJobs
+   * @param manager
+   */
+  public static void jobGarbageCollection(final Set<String> toBePurgedJobs,
```

Comment:
I feel purgeExpiredJobs has a more complete logic to process if the removal fails. Why not using it directly. In this case, you won't need 2 lists in the event object. You can merge the job with no config to the expired job for purging.
Or is there any conflict logic there?

Reply:
They are different. Purged expiredJobs are for the jobs that exist in the queue and they have workflow config. However, this is a general method to delete the expired jobs whether they have workflow config or not, and also whether the jobs are associated with a queue or regular workflow. Also, expired jobs are for the queue that is currently running. If the queue has been deleted and the job didn't get deleted for some reason, then this part will cover it and does garbage collection.